### PR TITLE
[WIP] Document compatibility tests

### PR DESCRIPTION
**Pull request by weralabaj**
Connects to https://github.com/Particular/docs.particular.net/issues/1247

### CODE REVIEW

File: nservicebus/testing/cross-version-testing.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+title: Cross-version compatibility testing with NServiceBus
+summary: Develop service layers and long-running processes using test-driven development.
+reviewed: 2016-06-06
+related:
+ - samples/unit-testing
+---
+
+In order to ensure that NServiceBus upgrades are safe and that endpoints in various versions can communicate with each other, it is necessary to extensively test multiple aspects of cross-version compatibility. As much as possible, those tests are automated. Some of them are available publicly in the [Particular/EndToEnd](https://github.com/Particular/EndToEnd/) repository.
+
+Those tests might be useful for NServiceBus users in order to verify that the framework works correctly in an unusual scenario, for example when using an unusually complex data structure, in order to understand in detail how specific feature works, or to reproduce an issue they experienced. The provided code can be run on a local machine, the pre-requisites are listed in the [Readme](https://github.com/Particular/EndToEnd/blob/master/README.md) file in the repository.
```
Comment (HEskandari) [inline nservicebus/testing/cross-version-testing.md]: I don't like this paragraph as a whole. It reads as if there's not enough work is done to test the edge-cases and the user will be on their own, which obviously is not the case. Do we need to justify why releasing this source code to public is a good thing or how it can be used? isn't that already implied?

Comment (SimonCropp) [inline nservicebus/testing/cross-version-testing.md]: > Those tests might be useful for NServiceBus users in order to verify that the framework works correctly in an unusual scenario, for example when using an unusually complex data structure, in order to understand in detail how specific feature works, or to reproduce an issue they experienced.

I would prefer some runnable code that shouw them how they can verify this stuff between versions. it would take a fair bit of effort to groc our EndToEnd enough to make them useful inside a endpoint

Comment (weralabaj) [inline nservicebus/testing/cross-version-testing.md]: So @HEskandari @SimonCropp would it be better just to skip the whole justification and start with running on local?

> I would prefer some runnable code that shouw them how they can verify this stuff between versions.

Not sure what you have in mind? Snippets?

Comment (SimonCropp) [inline nservicebus/testing/cross-version-testing.md]: well i think u need to phrase this in terms of what they want to do and what we want to achieve

one scenario is people wanting to evolve saga data/messages moving forward. how would u write tests for this? and repeat those test over multiple iterations. while there is much code in out endtoend we could use to do that , leaving that to the user to work out how to do that is a big ask

another scenario is someone building a persister/transport/serializer. in that case there is a much higher amount of code that could be reused from endtoend

the 3rd point we want to get across is that "we do backwards compat test and we take it seriously". this is more pitched at the enterprise architect level and is almost marketing speak.

Comment (weralabaj) [inline nservicebus/testing/cross-version-testing.md]: the 3rd will be addressed by a a story anyway, so I think focusing on option 2 makes the most sense at the moment. the 1st - not sure how often anybody would use it? also probably would be better to describe it, when we isolate the re-usable parts (generating versions pairs, pulling nugets, etc.) into separate libraries?

Comment (SimonCropp) [inline nservicebus/testing/cross-version-testing.md]: >  when we isolate the re-usable parts (generating versions pairs, pulling nugets, etc.) into separate libraries?

yes. in fact this has already been happening. samples and snippets regularly drive the design moving forwrd. for example most of the scripting snippets (http://docs.particular.net/nservicebus/msmq/operations-scripting) i would eventually like to see them as reusable libs.

Comment (weralabaj) [inline nservicebus/testing/cross-version-testing.md]: Might be worth opening an issue for it and getting a TF to move this forward faster?

**Comment by weralabaj** [comment]
@SimonCropp @HEskandari As per our discussion, I connected this issue to the documenting backward compat policy. I don't think that should be a blocker for closing https://github.com/Particular/PlatformDevelopment/issues/737.

**Comment by tmasternak** [comment]
We closing this. The content should be a good starting point to TF that will create backwards compatibility policy document.
